### PR TITLE
Container SBOM

### PR DESCRIPTION
**Pull request by tommyknows**
First commit is currently opened in https://github.com/snyk/cli-extension-sbom/pull/40.
Technically I think we could have a PR per commit. Not sure how much sense that makes, but I'm happy to open any number and combinations of PRs. Have a look at the commit history and let me know how I should split it up. 

---

**feat: single depGraph with subject**

This commit adds functionality to the sbom service, allowing users to
pass in a single depGraph *and* a subject that will be honored.

To achieve this, we simply send the multi-depgraph body with only one
depGraph, and the subject with it.

Additionally, I've switched from using raw strings to construct the
body-JSON to using `json.RawMessage` and proper types / structs. This
also means that I've had to make the `Name` and `Version` fields of the
`subject` type public. And to make the subject's usage easier in
external packages, I've made the `Subject` type itself public as well.

`json.RawMessage` is nice because it can be used to tell Go that a
certain field is already raw JSON, and has never been fully decoded into
a custom type. As such, when marshaling JSON, that `RawMessage` is
simply added to the rest of the document without any changes.

**chore: port depgraph workflow**

This commit simply copies the original depgraph workflow from the
`go-application-framework` into this repository.

Nothing has been modified apart from the filenames.

**refactor: depGraph extraction**

This commit extracts the depGraph extraction into it's own function, and
starts using a regex to extract the depGraph from the CLI's output.
Additionally, it adds more thorough testing with actual data extracted
from a CLI run.

**chore: introduce flag package**

The flag package abstracts "generic" string or boolean flags, which
makes typed access to specific flags easier.

Currently, nothing uses this new flag package, but that will change in
subsequent commits.

**refactor: abstract depgraph opensource implementation**

This commit abstracts away the opensource-specific parts of the depgraph
subcommand. The opensource-specific parts are now in
`internal/opensource/depgraph`, while the generic depgraph package stays
at `pkg/depgraph` so that it can be used by external modules as well.

Additionally, it adds a top-level package that exposes that internal
`depgraph` subcommand so that it can be registered at the CLI engine.

**refactor: abstract sbom opensource implementation**

This commit abstracts away the opensource-specific parts from the sbom
subcommand. The opensource-specific parts are now in
`internal/opensource/sbom`, while the generic sbom package stays at
`pkg/sbom` so that it can be used by external modules as well.

**test: more testing for SBOM**

This commit adds integration tests for the opensource SBOM command, as
well as a test package for external users to do the same.

The integration test for the SBOM command uses a mocked legacy CLI and a
mocked SBOM conversion service, but ensures that the `sbom` and
`depgraph` commands work together appropriately.

The mocks have been moved to the `test` package as well, as it was the
most natural place to put it (keep in mind that the SBOM conversion
service mock will have external users as well).

**fix: use Go 1.20 to build & test**

With Go 1.19, I was running into funky
[bugs](https://app.circleci.com/pipelines/github/snyk/cli-extension-sbom/138/workflows/b45a36f4-b767-48bf-b19f-5dd4a99db391)
probably related to the generics implementation?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -32,7 +32,7 @@ jobs:
     resource_class: medium
     executor:
       name: go/default
-      tag: '1.19'
+      tag: "1.20"
```
Comment (tommyknows) [inline .circleci/config.yml:35]: This obviously also needs a bump in the CLI....I'm currently checking with Hammerhead on this.

File: test/integration_test.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		config            map[string]any
+	}
+	tcs := []testCase{{
+		command:           "sbom",
+		scanResult:        openSourceScanOutput,
+		expectedDepGraphs: openSourceDepGraph,
+		config: map[string]any{
+			// doesn't matter much.
+			"targetDirectory": "./testdata/container_scan_output.txt",
+		},
```
Comment (tommyknows) [inline test/integration_test.go:41]: Obviously we could also test a combination of different flags & whatnot, but as the underlying scan is mocked away, I'm not sure if that really makes sense :)

File: .circleci/config.yml (hunk 1 above)
Comment (tommyknows) [reply .circleci/config.yml:35]: Apparently the CLI's already being built with Go 1.20 👍

**Comment by tommyknows** [comment]
After talking to the team, it seems like the goal would be to move the `depgraph` implementation to a separate repository still.
There's a couple of ways to do that, without creating a cyclic dependency between the repos. 

In the current version of the code, we have the following dependencies:

```mermaid
graph TD
sbom-os[pkg/sbom/opensource] --> sbom-main[pkg/sbom]
sbom-main --> flags[pkg/flag]
sbom-os --> flags
dg-main --> flags
dg-os --> flags
sbom-co[github.com/snyk/container-cli/sbom] --> sbom-main
dg-os[pkg/depgraph/opensource] --> dg-main[pkg/depgraph]
dg-co[github.com/snyk/container-cli/depgraph] --> dg-main
sbom-co --> flags
dg-co --> flags
os-test[test_test] --> sbom-os
os-test --> dg-os
os-test --> main-test[test]
co-test[github.com/snyk/container-cli/test_test] --> sbom-co
co-test --> dg-co
co-test --> main-test
```

(`internal` packages, other modules and unit-test-files omitted for simplicity)

Or, put in words:
- pretty much everything depends on `pkg/flag`
- `pkg/sbom/opensource` and `github.com/snyk/container-cli/sbom` both depend on `pkg/sbom`
- `pkg/depgraph/opensource` and `github.com/snyk/container-cli/depgraph` both depend on `pkg/depgraph`
- the integration test packages (the ones with `_test` suffix) depend on their respective `sbom` and `depgraph` implementations, and the `test` library.


Some proposals to split that up:
## by "responsibility"
- have a repo that contains both `pkg/sbom` and `pkg/depgraph` (the "abstractions") including the `test` and `flag` package
- create a repo for the opensource-specific implementations; `pkg/sbom/opensource` and `pkg/depgraph/opensource`, that includes the integration-tests (using the `test` package from above).
- have another repo for the container-specific implementations; `github.com/snyk/container-cli/sbom` and `github.com/snyk/container-cli/depgraph`, that also includes integration tests.

## by subcommand
- have a repo for packages that are used across both commands, e.g. currently the `pkg/flag`. This could also live in hammerheads `go-application-framework` potentially.
- a repo for `pkg/sbom` and `pkg/sbom/opensource`
- another repo for `pkg/depgraph` and `pkg/depgraph/opensource`
- another repo for the container-parts
- the `test` package also needs to go somewhere, and as all `pkg/sbom/opensource`, `pkg/depgraph/opensource` and the respective container-packages depend on it, it's probably also a separate repo.

## by package (lol)
give up, create a module / git repository per go package

----

I would very clearly vote for the first proposal, as the others are just very annoying to work with and introduce a lot of repositories....but in any of these proposals, `container` can do what they want, so I don't care :)

**Comment by mcombuechen** [comment]
Closing this since `snyk container sbom` was shipped meanwhile.

**Pull request closed without merging**